### PR TITLE
Check whether the barcode whitelist matches the data

### PR DESCRIPTION
**Pull request by mourisl**
Also added the option "--skip-barcode-check" to skip this step.

### CODE REVIEW

File: src/chromap.cc [hunk 1]
```diff
@@ -5968,8 +5976,8 @@ void ChromapDriver::ParseArgsAndRun(int argc, char *argv[]) {
       //("p,matrix-output-prefix", "Prefix of matrix output files",
       // cxxopts::value<std::string>(), "FILE")
       ("output-mappings-not-in-whitelist",
-       "Output mappings with barcode not in the whitelist")(
-          "chr-order", "custom chromsome order", cxxopts::value<std::string>(),
+       "Output mappings with barcode not in the whitelist")
```
Comment (haowenz) [inline src/chromap.cc:5979]: No need to change the code here?

File: src/chromap.h [hunk 2]
```diff
@@ -681,6 +683,8 @@ class Chromap {
   uint32_t barcode_length_ = 0;
   khash_t(k64_seq) * barcode_histogram_;
   khash_t(k64_seq) * barcode_index_table_;
+	bool skip_barcode_check_ = false ;
```
Comment (haowenz) [inline src/chromap.h:686]: Can we fix the leading space?

File: src/chromap.cc [hunk 3]
```diff
@@ -5999,7 +6007,9 @@ void ChromapDriver::ParseArgsAndRun(int argc, char *argv[]) {
       "drop-repetitive-reads",
       "Drop reads with too many best mappings [500000]", cxxopts::value<int>(),
       "INT")("allocate-multi-mappings", "Allocate multi-mappings")(
-      "PAF", "Output mappings in PAF format (only for test)");
+      "PAF", "Output mappings in PAF format (only for test)")
+			("skip-barcode-check", "Do not check whether too few barcodes are in the whitelist")
```
Comment (haowenz) [inline src/chromap.cc:6011]: Also fix the leading space here.

File: src/chromap.cc [hunk 4]
```diff
@@ -749,6 +749,13 @@ void Chromap<MappingRecord>::ComputeBarcodeAbundance(
           ++num_sample_barcodes_;
         }
       }
+
+			if (!skip_barcode_check_
```
Comment (haowenz) [inline src/chromap.cc:753]: This should be okay. Can we also fix the leading space here?

File: src/chromap.cc [hunk 5]
```diff
@@ -749,6 +749,13 @@ void Chromap<MappingRecord>::ComputeBarcodeAbundance(
           ++num_sample_barcodes_;
         }
       }
+
+			if (!skip_barcode_check_  
+					&& num_sample_barcodes_ * 20 < num_loaded_barcodes) {
+				// Since num_loaded_pairs is a constant, this if is actuaclly only effective in the first iteration
+				Chromap<>::ExitWithMessage("Less than 5\% barcodes can be found or corrected based on the barcode whitelist.\nPlease check whether the barcode whitelist matches the data, e.g. length, reverse-complement. If this is a false positive warning, please run Chromap with the option --skip-barcode-check.");
```
Comment (haowenz) [inline src/chromap.cc:756]: Not sure if "5\%" is necessary or just "5%" is fine.

**Review by haowenz — CHANGES_REQUESTED** [review]
Can you fix the format of the code? Especially the leading space for some lines.

**Comment by mourisl** [comment]
The leading tabs should be spaces now.

**Review by haowenz — APPROVED** [review]

**Pull request merged**